src/protocol/boolean/bitwise_sum.rs

```diff
@@ -0,0 +1,212 @@
+
+        // Step 2. `[d_0] = [a_0] + [b_0] - 2[c_1]`
+        // The paper refers `[c]_b` as `([c_1],...[c_l])`; the starting index is 1;
+        // therefore, `[c_1]` is the first element, `c[0]`, in the code.
+        let mut d = Vec::with_capacity(l);
+        d.push(a[0] + b[0] - (c[0] * F::from(2)));
+
+        // Step 3. `[d_l] = [c_l]`
+        // Step 4. for `i=1..l-1`, `[d_i] = [a_i] + [b_i] + [c_i] - 2[c_(i+1)]`
+        //
+        // In the paper, the output of BIT-ADD is `l + 1` long, where `l` is
+        // the length of the inputs `a` and `b`. So, if we are working with a
+        // 32-bit long field, the output of this protocol will be 33-bit long
+        // bitwise shares.
+        for i in 1..l {
+            d.push(a[i] + b[i] + c[i - 1] - (c[i] * F::from(2)));
+        }
+        d.push(c[l - 1]);
+
+        Ok(d)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    Carries,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::Carries => "carries",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::BitwiseSum;
+    use crate::{
+        error::BoxError,
+        ff::{Field, Fp31, Fp32BitPrime, Int},
+        protocol::{QueryId, RecordId},
+        secret_sharing::Replicated,
+        test_fixture::{make_contexts, make_world, share, validate_and_reconstruct, TestWorld},
+    };
+    use futures::future::try_join_all;
+    use rand::{distributions::Standard, prelude::Distribution, rngs::mock::StepRng, RngCore};
+
+    /// From `Vec<[Replicated<F>; 3]>`, create `Vec<Replicated<F>>` taking the `i`'th share per row
+    fn transpose<F: Field>(x: &[[Replicated<F>; 3]], i: usize) -> Vec<Replicated<F>> {
+        x.iter().map(|&x| x[i]).collect::<Vec<_>>()
+    }
+
+    /// Take a field value `x` and turn them into replicated bitwise sharings of three
+    fn shared_bits<F: Field, R: RngCore>(
+        x: F,
+        bit_len: u32,
+        rand: &mut R,
+    ) -> Vec<[Replicated<F>; 3]>
+    where
+        Standard: Distribution<F>,
+    {
+        let x = x.as_u128();
+        (0..bit_len)
+            .map(|i| share(F::from(x >> i & 1), rand))
+            .collect::<Vec<_>>()
+    }
+
+    /// Take a slice of bits in `{0,1} ⊆ F_p`, and reconstruct the integer in `F_p`
+    fn bits_to_field<F: Field>(x: &[F]) -> F {
+        #[allow(clippy::cast_possible_truncation)]
+        let v = x
+            .iter()
+            .enumerate()
+            .fold(0, |acc, (i, &b)| acc + 2_u128.pow(i as u32) * b.as_u128());
+        F::from(v)
+    }
+
+    #[allow(clippy::many_single_char_names)]
+    async fn bitwise_sum<F: Field>(a: F, b: F) -> Result<Vec<F>, BoxError>
+    where
+        Standard: Distribution<F>,
+    {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<F>(&world);
+        let mut rand = StepRng::new(1, 1);
```

Comment:
Creating the world and contexts within this function means you're doing so once for every test. That's a lot of times. Maybe move this out and pass them in?

Reply:
If I move out `ctx`, which is an array of 3 context instances, out, I will need to `narrow()` them each time I call `bitwise_sum` helper function. I think the cost is almost the same, and easier to read this way.

Reply:
I don’t think those compare in cost at all. Narrowing the context makes a single struct. Creating the world takes a long time because of the key exchange to set up the PRSS